### PR TITLE
ecs-systems-locator-490 - Refactor ECS system to use the locator pattern

### PR DESCRIPTION
**Pull request by raffclar**

### CODE REVIEW

File: src/Locator.h [hunk 1]
```diff
@@ -9,10 +9,15 @@
 
 #pragma once
 
+#include <ECS/Systems/PathfindingSystem.h>
```
Comment (bwrsandman) [inline src/Locator.h:12]: Don't use angle brackets for non system includes

File: src/ECS/Systems/LivingActionSystem.h [hunk 2]
```diff
@@ -13,24 +13,38 @@
...
-	bool VillagerCallExitState(components::LivingAction& action, components::LivingAction::Index index) const;
-	int VillagerCallOutOfAnimation(components::LivingAction& action, components::LivingAction::Index index) const;
-	bool VillagerCallValidate(components::LivingAction& action, components::LivingAction::Index index) const;
-
-private:
-	LivingActionSystem();
+	                            VillagerStates dst) const override;
+	bool VillagerCallExitState(components::LivingAction& action, components::LivingAction::Index index) const override;
+	int VillagerCallOutOfAnimation(components::LivingAction& action, components::LivingAction::Index index) const override;
+	bool VillagerCallValidate(components::LivingAction& action, components::LivingAction::Index index) const override;
```
Comment (bwrsandman) [inline src/ECS/Systems/LivingActionSystem.h:48]: Private functions do not need to be virtual

File: src/ECS/Systems/CameraBookmarkSystem.h [hunk 3]
```diff
@@ -18,21 +18,28 @@
...
+	void Update(const std::chrono::microseconds& dt) const override;
 
-	void SetBookmark(uint8_t index, const glm::vec3& position) const;
-	void ClearBookmark(uint8_t index) const;
+	[[nodiscard]] const std::array<entt::entity, 8>& GetBookmarks() const override { return _bookmarks; }
 
-private:
-	CameraBookmarkSystem();
+	void SetBookmark(uint8_t index, const glm::vec3& position) const override;
+	void ClearBookmark(uint8_t index) const override;
```
Comment (bwrsandman) [inline src/ECS/Systems/CameraBookmarkSystem.h:40]: Private functions don't need to be virtual

**Review by bwrsandman — CHANGES_REQUESTED** [review]
Not sure how but your changes broke Enums that are named None. Probably because a None got a value somewhere.
This can probably be fixed by reordering includes or renaming None to `NONE` (Not a fan of renaming)

File: src/ECS/Systems/CameraBookmarkSystem.h (hunk 3 above)
Comment (raffclar) [reply src/ECS/Systems/CameraBookmarkSystem.h:40]: These are public before the refactor.

File: src/Locator.h (hunk 1 above)
Comment (raffclar) [reply src/Locator.h:12]: Nice catch

File: src/ECS/Systems/LivingActionSystem.h (hunk 2 above)
Comment (raffclar) [reply src/ECS/Systems/LivingActionSystem.h:48]: Same here. These are all public before the refactor

File: src/Resources/MeshId.h [hunk 4]
```diff
@@ -9,6 +9,8 @@
 
 #pragma once
 
+#include <entt/entt.hpp>
```
Comment (bwrsandman) [inline src/Resources/MeshId.h:12]: entt/core/hashed_string.hpp

File: src/Locator.h [hunk 5]
```diff
@@ -11,17 +11,35 @@
...
+
+namespace ecs::systems
+{
+class RenderingSystemInterface;
+class DynamicsSystemInterface;
+class CameraBookmarkSystemInterface;
+class LivingActionSystemInterface;
+class TownSystemInterface;
+class PathfindingSystemInterface;
+} // namespace ecs::systems
```
Comment (bwrsandman) [inline src/Locator.h:32]: Awesome!

**Review by bwrsandman — CHANGES_REQUESTED** [review]
Is it a good idea to have the Abstract base classes in the same headers as the default provided definitions?

File: src/Resources/MeshId.h (hunk 4 above)
Comment (raffclar) [reply src/Resources/MeshId.h:12]: > Is it a good idea to have the Abstract base classes in the same headers as the default provided definitions?

No it isn't. We don't have multiple implementations and we have forward declarations now so there's no performance to be gained.

Might take a bit of work to do but probably best to do it now before we have 100 systems.

Comment (raffclar) [reply src/Resources/MeshId.h]: > entt/core/hashed_string.hpp

Thanks

**Review by bwrsandman — APPROVED** [review]

File: src/Resources/MeshId.h (hunk 4 above)
Comment (raffclar) [reply src/Resources/MeshId.h]: > > Is it a good idea to have the Abstract base classes in the same headers as the default provided definitions?
> 
> No it isn't. We don't have multiple implementations and we have forward declarations now so there's no performance to be gained.
> 
> Might take a bit of work to do but probably best to do it now before we have 100 systems.

@bwrsandman #492

**Pull request merged**